### PR TITLE
distribute framework: add planner

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/46258

Problem Summary:

### What is changed and how it works?
- logical plan -> task meta -> physical plan -> subtaskmetas -> operators.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/framework/planner/plan.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+	// integrate with current distribute framework
+	SessionCtx sessionctx.Context
+	TaskKey    string
+	TaskType   string
+	ThreadCnt  int
+
+	// PreviousSubtaskMetas is a list of subtask metas from previous step.
+	// We can remove this field if we find a better way to pass the result between steps.
+	PreviousSubtaskMetas [][]byte
```
Comment (ywqzzy) [inline disttask/framework/planner/plan.go:35]: hold task_table as member?

File: disttask/importinto/planner.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+	p.ChunkMap = taskMeta.ChunkMap
+	return nil
+}
+
+// ToPhysicalPlan converts the logical plan to physical plan.
+func (p *LogicalPlan) ToPhysicalPlan(planCtx planner.PlanCtx) (*planner.PhysicalPlan, error) {
+	physicalPlan := &planner.PhysicalPlan{}
+	inputStreams := make([]planner.StreamSpec, 0)
+	// physical plan only needs to be generated once.
+	// However, our current implementation requires generating it for each step.
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply disttask/importinto/planner.go:77]: if we support pass result from one step two another step, then we only need generated the initial physical plan, all runtime arguments should be define in input stream.

File: disttask/importinto/planner.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+	_ planner.OperatorSpec = &ImportSpec{}
+	_ planner.OperatorSpec = &PostProcessSpec{}
+)
+
+// LogicalPlan represents a logical plan for import into.
+type LogicalPlan struct {
+	JobID             int64
+	Plan              importer.Plan
+	Stmt              string
+	EligibleInstances []*infosync.ServerInfo
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply disttask/importinto/planner.go:42]: https://github.com/pingcap/tidb/blob/2564090c660157608710a150f4884e7891f7f0e3/disttask/importinto/job.go#L185

File: disttask/framework/planner/plan.go (hunk 1 above)
Comment (GMHDBJD) [reply disttask/framework/planner/plan.go:35]: Using PreviousSubtaskMetas is to express that PreviousSubtaskMetas is the result of the previous stage. If we use task_table, our planner will be coupled with the system table again, although adding a new interface would be more convenient. 🤔

File: disttask/framework/planner/plan.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+type PlanCtx struct {
+	Ctx context.Context
+
+	// integrate with current distribute framework
+	SessionCtx sessionctx.Context
+	TaskKey    string
+	TaskType   string
+	ThreadCnt  int
+
+	// PreviousSubtaskMetas is a list of subtask metas from previous step.
```
Comment (Benjamin2037) [inline disttask/framework/planner/plan.go:33]: we should plan need care about previous Subtask metas?

Comment (GMHDBJD) [reply disttask/framework/planner/plan.go:33]: We need to pass the result from one step to next step, currently we can only pass it by subtask metas.

File: disttask/framework/planner/plan.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	ColumnTypes []byte
+	Streams     []StreamSpec
+}
+
+// OutputSpec is the specification of an output.
+type OutputSpec struct {
+	Streams []StreamSpec
+}
+
+// StreamSpec is the specification of a stream.
```
Comment (ywqzzy) [inline disttask/framework/planner/plan.go:98]: The name is a little bit confusing, can we have a better name?

File: disttask/framework/planner/planner_test.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/pingcap/tidb/disttask/framework/planner"
+	"github.com/pingcap/tidb/disttask/framework/storage"
+	"github.com/pingcap/tidb/testkit"
+	"github.com/stretchr/testify/require"
+	"github.com/tikv/client-go/v2/util"
+)
+
+type mockPlan struct{}
+
+func (mockPlan) ToTaskMeta() ([]byte, error) {
```
Comment (ywqzzy) [inline disttask/framework/planner/planner_test.go:32]: Maybe we can use gen_mock introduced in https://github.com/pingcap/tidb/pull/46392

File: disttask/framework/planner/planner_test.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func (mockPlan) ToPhysicalPlan(planner.PlanCtx) (*planner.PhysicalPlan, error) {
+	return &planner.PhysicalPlan{}, nil
+}
+
+func TestPlanner(t *testing.T) {
+	ctx := context.Background()
+	store := testkit.CreateMockStore(t)
+	gtk := testkit.NewTestKit(t, store)
```
Comment (ywqzzy) [inline disttask/framework/planner/planner_test.go:47]: Seems useless? Bootstrap a testkit cost some resource, better not add it.

File: disttask/framework/planner/plan.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+// To integrate with current distribute framework, the flow becomes:
+// logical plan -> task meta -> physical plan -> subtaskmetas -> operators.
+type LogicalPlan interface {
+	ToTaskMeta() ([]byte, error)
+	FromTaskMeta([]byte) error
+	ToPhysicalPlan(PlanCtx) (*PhysicalPlan, error)
+}
+
+// PhysicalPlan is a DAG of processors in distribute framework.
+type PhysicalPlan struct {
```
Comment (ywqzzy) [inline disttask/framework/planner/plan.go:49]: Can we add  more detailed comments to show the arch of the planner?

**Review by ywqzzy — APPROVED** [review]
LGTM

File: disttask/framework/planner/planner_test.go (hunk 7 above)
Comment (GMHDBJD) [reply disttask/framework/planner/planner_test.go:38]: we need testkit to submit a global task like what we do in handle_test https://github.com/pingcap/tidb/blob/7e66761047b01b62c4097892be14555f83a3c3b3/disttask/framework/handle/handle_test.go#L34

File: disttask/framework/planner/planner.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+package planner
+
+import "github.com/pingcap/tidb/disttask/framework/storage"
+
+// Planner represents a distribute plan planner.
+type Planner struct{}
+
+// Run runs the distribute plan.
+func (*Planner) Run(planCtx PlanCtx, plan LogicalPlan) (int64, error) {
```
Comment (tangenta) [inline disttask/framework/planner/planner.go:23]: No usage?

**Review by D3Hunter — COMMENTED** [review]
5/13 review later

**Review by tangenta — APPROVED** [review]

**Pull request merged**